panic: runtime error: invalid memory address or nil pointer dereference

macos   10.15.7
CLI version: 1.9.1
Runtime version: 1.9.3

dapr run --app-id dicom-service --app-port 3000 --app-protocol http --dapr-http-port 3001  --dapr-http-max-request-size 1024 --dapr-grpc-port 3002  node .

curl http://localhost:3001/v1.0/invoke/dicom-service/method/series
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x44caea9]

goroutine 162 [running]:
github.com/dapr/dapr/pkg/messaging.(*directMessaging).getRemoteApp(0xc00017a0c0, {0xc000ae4be0?, 0x8020102?})
	/Users/runner/work/dapr/dapr/pkg/messaging/direct_messaging.go:309 +0xc9
github.com/dapr/dapr/pkg/messaging.(*directMessaging).Invoke(0xc00017a0c0, {0x5cf1208, 0xc000f08600}, {0xc000ae4be0?, 0x203000?}, 0xc00064b620?)
	/Users/runner/work/dapr/dapr/pkg/messaging/direct_messaging.go:126 +0x56
github.com/dapr/dapr/pkg/http.(*api).onDirectMessage.func1({0x5cf1208, 0xc000f08600})
	/Users/runner/work/dapr/dapr/pkg/http/api.go:1390 +0xba
github.com/dapr/dapr/pkg/resiliency.(*NoOp).EndpointPolicy.func1(0x53dd0f8?)
	/Users/runner/work/dapr/dapr/pkg/resiliency/noop.go:36 +0x27
github.com/dapr/dapr/pkg/http.(*api).onDirectMessage(0xc0006da000, 0xc000f08600)
	/Users/runner/work/dapr/dapr/pkg/http/api.go:1389 +0x6d6
github.com/fasthttp/router.(*Router).Handler(0xc00039a460, 0xc000f08600)
	/Users/runner/go/pkg/mod/github.com/fasthttp/router@v1.4.12/router.go:427 +0x871
github.com/dapr/dapr/pkg/diagnostics.(*httpMetrics).FastHTTPMiddleware.func1(0xc000f08600)
	/Users/runner/work/dapr/dapr/pkg/diagnostics/http_monitoring.go:227 +0x13b
github.com/dapr/dapr/pkg/diagnostics.HTTPTraceMiddleware.func1(0xc000f08600)
	/Users/runner/work/dapr/dapr/pkg/diagnostics/http_tracing.go:56 +0x1ad
github.com/valyala/fasthttp.(*Server).serveConn(0xc0006d66c0, {0x5d04c10?, 0xc00064e300})
	/Users/runner/go/pkg/mod/github.com/valyala/fasthttp@v1.40.0/server.go:2311 +0x1268
github.com/valyala/fasthttp.(*workerPool).workerFunc(0xc00055cd20, 0xc0006faa20)
	/Users/runner/go/pkg/mod/github.com/valyala/fasthttp@v1.40.0/workerpool.go:224 +0xa9
github.com/valyala/fasthttp.(*workerPool).getCh.func1()
	/Users/runner/go/pkg/mod/github.com/valyala/fasthttp@v1.40.0/workerpool.go:196 +0x38
created by github.com/valyala/fasthttp.(*workerPool).getCh
	/Users/runner/go/pkg/mod/github.com/valyala/fasthttp@v1.40.0/workerpool.go:195 +0x1b0
❌  The daprd process exited with error code: exit status 2

@ItalyPaleAle  thanks for your profession help.The problem is just because my consul serivce crashed down but I didn't find it.Hope my logs can help you.

Here is the logs from startup to crash .

jy@jys-MacBook-Pro dicom-service % dapr run --app-id test --dapr-http-port 3501
WARNING: no application command found.
ℹ️  Starting Dapr with id test. HTTP Port: 3501. gRPC Port: 58449
ℹ️  Checking if Dapr sidecar is listening on HTTP port 3501
INFO[0000] starting Dapr Runtime -- version 1.8.1 -- commit f2583aa5cf166d83bedc129dfbbd720f100d73d2  app_id=test instance=jys-MacBook-Pro scope=dapr.runtime type=log ver=1.8.1
INFO[0000] log level set to: info                        app_id=test instance=jys-MacBook-Pro scope=dapr.runtime type=log ver=1.8.1
INFO[0000] metrics server started on :58450/             app_id=test instance=jys-MacBook-Pro scope=dapr.metrics type=log ver=1.8.1
INFO[0000] standalone mode configured                    app_id=test instance=jys-MacBook-Pro scope=dapr.runtime type=log ver=1.8.1
INFO[0000] app id: test                                  app_id=test instance=jys-MacBook-Pro scope=dapr.runtime type=log ver=1.8.1
INFO[0000] mTLS is disabled. Skipping certificate request and tls validation  app_id=test instance=jys-MacBook-Pro scope=dapr.runtime type=log ver=1.8.1
ERRO[0000] failed to initialize name resolution resolver consul: failed to register consul service: Put "http://127.0.0.1:8500/v1/agent/service/register": dial tcp 127.0.0.1:8500: connect: connection refused  app_id=test instance=jys-MacBook-Pro scope=dapr.runtime type=log ver=1.8.1
WARN[0000] failed to init name resolution: failed to register consul service: Put "http://127.0.0.1:8500/v1/agent/service/register": dial tcp 127.0.0.1:8500: connect: connection refused  app_id=test instance=jys-MacBook-Pro scope=dapr.runtime type=log ver=1.8.1
INFO[0000] loading components                            app_id=test instance=jys-MacBook-Pro scope=dapr.runtime type=log ver=1.8.1
INFO[0000] component loaded. name: pubsub, type: pubsub.redis/v1  app_id=test instance=jys-MacBook-Pro scope=dapr.runtime type=log ver=1.8.1
INFO[0000] waiting for all outstanding components to be processed  app_id=test instance=jys-MacBook-Pro scope=dapr.runtime type=log ver=1.8.1
INFO[0000] detected actor state store: statestore        app_id=test instance=jys-MacBook-Pro scope=dapr.runtime type=log ver=1.8.1
INFO[0000] component loaded. name: statestore, type: state.redis/v1  app_id=test instance=jys-MacBook-Pro scope=dapr.runtime type=log ver=1.8.1
INFO[0000] all outstanding components processed          app_id=test instance=jys-MacBook-Pro scope=dapr.runtime type=log ver=1.8.1
INFO[0000] gRPC proxy enabled                            app_id=test instance=jys-MacBook-Pro scope=dapr.runtime type=log ver=1.8.1
INFO[0000] enabled gRPC tracing middleware               app_id=test instance=jys-MacBook-Pro scope=dapr.runtime.grpc.api type=log ver=1.8.1
INFO[0000] enabled gRPC metrics middleware               app_id=test instance=jys-MacBook-Pro scope=dapr.runtime.grpc.api type=log ver=1.8.1
INFO[0000] API gRPC server is running on port 58449      app_id=test instance=jys-MacBook-Pro scope=dapr.runtime type=log ver=1.8.1
INFO[0000] enabled metrics http middleware               app_id=test instance=jys-MacBook-Pro scope=dapr.runtime.http type=log ver=1.8.1
INFO[0000] enabled tracing http middleware               app_id=test instance=jys-MacBook-Pro scope=dapr.runtime.http type=log ver=1.8.1
INFO[0000] http server is running on port 3501           app_id=test instance=jys-MacBook-Pro scope=dapr.runtime type=log ver=1.8.1
INFO[0000] The request body size parameter is: 4         app_id=test instance=jys-MacBook-Pro scope=dapr.runtime type=log ver=1.8.1
INFO[0000] enabled gRPC tracing middleware               app_id=test instance=jys-MacBook-Pro scope=dapr.runtime.grpc.internal type=log ver=1.8.1
INFO[0000] enabled gRPC metrics middleware               app_id=test instance=jys-MacBook-Pro scope=dapr.runtime.grpc.internal type=log ver=1.8.1
INFO[0000] internal gRPC server is running on port 58454  app_id=test instance=jys-MacBook-Pro scope=dapr.runtime type=log ver=1.8.1
WARN[0000] app channel is not initialized. did you make sure to configure an app-port?  app_id=test instance=jys-MacBook-Pro scope=dapr.runtime type=log ver=1.8.1
INFO[0000] actor runtime started. actor idle timeout: 1h0m0s. actor scan interval: 30s  app_id=test instance=jys-MacBook-Pro scope=dapr.runtime.actor type=log ver=1.8.1
WARN[0000] app channel not initialized, make sure -app-port is specified if pubsub subscription is required  app_id=test instance=jys-MacBook-Pro scope=dapr.runtime type=log ver=1.8.1
WARN[0000] failed to read from bindings: app channel not initialized   app_id=test instance=jys-MacBook-Pro scope=dapr.runtime type=log ver=1.8.1
INFO[0000] dapr initialized. Status: Running. Init Elapsed 30.956ms  app_id=test instance=jys-MacBook-Pro scope=dapr.runtime type=log ver=1.8.1
ℹ️  Checking if Dapr sidecar is listening on GRPC port 58449
ℹ️  Dapr sidecar is up and running.
✅  You're up and running! Dapr logs will appear here.

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x22e1349]

goroutine 134 [running]:
github.com/dapr/dapr/pkg/messaging.(*directMessaging).getRemoteApp(0xc00066e8c0, {0xc000b45178?, 0x6000105?})
	/Users/runner/work/dapr/dapr/pkg/messaging/direct_messaging.go:310 +0xc9
github.com/dapr/dapr/pkg/messaging.(*directMessaging).Invoke(0xc00066e8c0, {0x56b8508, 0xc000eea580}, {0xc000b45178?, 0x203000?}, 0xc00111ee40?)
	/Users/runner/work/dapr/dapr/pkg/messaging/direct_messaging.go:126 +0x56
github.com/dapr/dapr/pkg/http.(*api).onDirectMessage.func1({0x56b8508, 0xc000eea580})
	/Users/runner/work/dapr/dapr/pkg/http/api.go:1392 +0xba
github.com/dapr/dapr/pkg/resiliency.(*NoOp).EndpointPolicy.func1(0x4eb9387?)
	/Users/runner/work/dapr/dapr/pkg/resiliency/noop.go:36 +0x27
github.com/dapr/dapr/pkg/http.(*api).onDirectMessage(0xc000cf1cc0, 0xc000eea580)
	/Users/runner/work/dapr/dapr/pkg/http/api.go:1391 +0x6e3
github.com/fasthttp/router.(*Router).Handler(0xc000dca1c0, 0xc000eea580)
	/Users/runner/go/pkg/mod/github.com/fasthttp/router@v1.3.8/router.go:427 +0x871
github.com/dapr/dapr/pkg/diagnostics.(*httpMetrics).FastHTTPMiddleware.func1(0xc000eea580)
	/Users/runner/work/dapr/dapr/pkg/diagnostics/http_monitoring.go:195 +0x13b
github.com/dapr/dapr/pkg/diagnostics.HTTPTraceMiddleware.func1(0xc000eea580)
	/Users/runner/work/dapr/dapr/pkg/diagnostics/http_tracing.go:56 +0xef
github.com/valyala/fasthttp.(*Server).serveConn(0xc0006c5800, {0x56c71e8?, 0xc000f041c8})
	/Users/runner/go/pkg/mod/github.com/valyala/fasthttp@v1.31.1-0.20211216042702-258a4c17b4f4/server.go:2293 +0x1268
github.com/valyala/fasthttp.(*workerPool).workerFunc(0xc000eaa0a0, 0xc000de07c0)
	/Users/runner/go/pkg/mod/github.com/valyala/fasthttp@v1.31.1-0.20211216042702-258a4c17b4f4/workerpool.go:223 +0xa9
github.com/valyala/fasthttp.(*workerPool).getCh.func1()
	/Users/runner/go/pkg/mod/github.com/valyala/fasthttp@v1.31.1-0.20211216042702-258a4c17b4f4/workerpool.go:195 +0x38
created by github.com/valyala/fasthttp.(*workerPool).getCh
	/Users/runner/go/pkg/mod/github.com/valyala/fasthttp@v1.31.1-0.20211216042702-258a4c17b4f4/workerpool.go:194 +0x1b0
❌  The daprd process exited with error code: exit status 2
ℹ️
terminated signal received: shutting down
❌  Error exiting Dapr: exit status 2

--- a/pkg/messaging/direct_messaging.go
+++ b/pkg/messaging/direct_messaging.go
@@ -137,12 +137,16 @@ func (d *directMessaging) Invoke(ctx context.Context, targetAppID string, req *i
 
 // requestAppIDAndNamespace takes an app id and returns the app id, namespace and error.
 func (d *directMessaging) requestAppIDAndNamespace(targetAppID string) (string, string, error) {
+	if targetAppID == "" {
+		return "", "", errors.New("app id is empty")
+	}
 	items := strings.Split(targetAppID, ".")
-	if len(items) == 1 {
+	switch len(items) {
+	case 1:
 		return targetAppID, d.namespace, nil
-	} else if len(items) == 2 {
+	case 2:
 		return items[0], items[1], nil
-	} else {
+	default:
 		return "", "", fmt.Errorf("invalid app id %s", targetAppID)
 	}
 }

--- a/pkg/messaging/direct_messaging.go
+++ b/pkg/messaging/direct_messaging.go
@@ -15,12 +15,13 @@ package messaging
 
 import (
 	"context"
+	"errors"
+	"fmt"
 	"os"
 	"strings"
 	"time"
 
 	"github.com/cenkalti/backoff/v4"
-	"github.com/pkg/errors"
 	"github.com/valyala/fasthttp"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/codes"
@@ -142,7 +143,7 @@ func (d *directMessaging) requestAppIDAndNamespace(targetAppID string) (string, 
 	} else if len(items) == 2 {
 		return items[0], items[1], nil
 	} else {
-		return "", "", errors.Errorf("invalid app id %s", targetAppID)
+		return "", "", fmt.Errorf("invalid app id %s", targetAppID)
 	}
 }
 
@@ -185,7 +186,7 @@ func (d *directMessaging) invokeWithRetry(
 			})
 			// To maintain consistency with the existing built-in retries, we do some transformations/error handling.
 			if retriesExhaustedPath {
-				return nil, errors.Errorf("failed to invoke target %s after %v retries. Error: %s", app.id, numRetries, err.Error())
+				return nil, fmt.Errorf("failed to invoke target %s after %v retries. Error: %s", app.id, numRetries, err.Error())
 			}
 
 			if nullifyResponsePath {
@@ -218,7 +219,7 @@ func (d *directMessaging) invokeWithRetry(
 		teardown(false)
 		return resp, err
 	}
-	return nil, errors.Errorf("failed to invoke target %s after %v retries", app.id, numRetries)
+	return nil, fmt.Errorf("failed to invoke target %s after %v retries", app.id, numRetries)
 }
 
 func (d *directMessaging) invokeLocal(ctx context.Context, req *invokev1.InvokeMethodRequest) (*invokev1.InvokeMethodResponse, error) {
@@ -331,6 +332,10 @@ func (d *directMessaging) getRemoteApp(appID string) (remoteApp, error) {
 	id, namespace, err := d.requestAppIDAndNamespace(appID)
 	if err != nil {
 		return remoteApp{}, err
+	}
+
+	if d.resolver == nil {
+		return remoteApp{}, errors.New("name resolver not initialized")
 	}
 
 	request := nr.ResolveRequest{ID: id, Namespace: namespace, Port: d.grpcPort}

--- a/pkg/messaging/direct_messaging_test.go
+++ b/pkg/messaging/direct_messaging_test.go
@@ -14,11 +14,14 @@ limitations under the License.
 package messaging
 
 import (
+	"errors"
+	"reflect"
 	"testing"
 
 	"github.com/stretchr/testify/assert"
 	"github.com/valyala/fasthttp"
 
+	nr "github.com/dapr/components-contrib/nameresolution"
 	invokev1 "github.com/dapr/dapr/pkg/messaging/v1"
 )
 
@@ -134,5 +137,92 @@ func TestKubernetesNamespace(t *testing.T) {
 		_, _, err := dm.requestAppIDAndNamespace(appID)
 
 		assert.Error(t, err)
+	})
+}
+
+type testResolver struct{}
+
+func (r testResolver) Init(metadata nr.Metadata) error {
+	// nop
+	return nil
+}
+func (r testResolver) ResolveID(req nr.ResolveRequest) (string, error) {
+	switch req.ID {
+	case "okapp":
+		return "12.34.56.78", nil
+	case "emptyapp":
+		return "", nil
+	case "failapp":
+		return "", errors.New("failed")
+	default:
+		panic("not a valid case")
+	}
+}
+
+func Test_directMessaging_getRemoteApp(t *testing.T) {
+	const defaultNamespace = "mynamespace"
+	type args struct {
+		appID string
+	}
+	tests := []struct {
+		name    string
+		args    args
+		want    remoteApp
+		wantErr bool
+	}{
+		{
+			name: "app with no namespace",
+			args: args{appID: "okapp"},
+			want: remoteApp{id: "okapp", namespace: defaultNamespace, address: "12.34.56.78"},
+		},
+		{
+			name: "app with namespace",
+			args: args{appID: "okapp.ns2"},
+			want: remoteApp{id: "okapp", namespace: "ns2", address: "12.34.56.78"},
+		},
+		{
+			name:    "empty appID",
+			args:    args{appID: ""},
+			wantErr: true,
+		},
+		{
+			name:    "invalid appID",
+			args:    args{appID: "not.valid.appID"},
+			wantErr: true,
+		},
+		{
+			name: "app doesn't exist",
+			args: args{appID: "emptyapp"},
+			want: remoteApp{id: "emptyapp", namespace: defaultNamespace, address: ""},
+		},
+		{
+			name:    "resolver errors",
+			args:    args{appID: "failapp"},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			d := &directMessaging{
+				grpcPort:  3500,
+				namespace: defaultNamespace,
+				resolver:  &testResolver{},
+			}
+			got, err := d.getRemoteApp(tt.args.appID)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("directMessaging.getRemoteApp() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("directMessaging.getRemoteApp() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+
+	t.Run("resolver is nil", func(t *testing.T) {
+		d := &directMessaging{}
+		_, err := d.getRemoteApp("foo")
+		_ = assert.Error(t, err) &&
+			assert.ErrorContains(t, err, "name resolver not initialized")
 	})
 }

--- a/pkg/messaging/direct_messaging_test.go
+++ b/pkg/messaging/direct_messaging_test.go
@@ -146,6 +146,7 @@ func (r testResolver) Init(metadata nr.Metadata) error {
 	// nop
 	return nil
 }
+
 func (r testResolver) ResolveID(req nr.ResolveRequest) (string, error) {
 	switch req.ID {
 	case "okapp":

--- a/pkg/runtime/initError.go
+++ b/pkg/runtime/initError.go
@@ -39,6 +39,10 @@ func (e *InitError) Error() string {
 	return fmt.Sprintf("[%s]: initialization error occurred for %s: %s", e.kind, e.entity, e.err)
 }
 
+func (e *InitError) Unwrap() error {
+	return e.err
+}
+
 // NewInitError returns an InitError wrapping an existing context error.
 func NewInitError(kind InitErrorKind, entity string, err error) *InitError {
 	return &InitError{

--- a/pkg/runtime/runtime.go
+++ b/pkg/runtime/runtime.go
@@ -2541,7 +2541,7 @@ func (a *DaprRuntime) shutdownOutputComponents() error {
 }
 
 func closeComponent(component any, logmsg string, merr *error) {
-	if closer, ok := component.(io.Closer); ok {
+	if closer, ok := component.(io.Closer); ok && closer != nil {
 		if err := closer.Close(); err != nil {
 			err = fmt.Errorf("error closing %s: %w", logmsg, err)
 			*merr = multierror.Append(*merr, err)

